### PR TITLE
kubeadm: delete bootstrap-kubelet.conf after TLS bootstrap

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
kubernetes/kubernetes#66482 enabled  `kubeadm join --discovery-file` with a discovery file with embedded credentials. This PR makes sure that kubeadm removes the bootstrap-kubelet.conf after completing the TLS bootstrap process, so copies of the above credentials won't be left around.

**Which issue(s) this PR fixes**:
Fixes #https://github.com/kubernetes/kubeadm/issues/1689

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
ACTION REQUIRED:
kubeadm now deletes the bootstrap-kubelet.conf file after TLS bootstrap
User relying on bootstrap-kubelet.conf should switch to kubelet.conf that contains node credentials
```

/sig cluster-lifecycle
/priority important-soon
/assign @timothysc
/assign @neolit123

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
i think this is mostly OK, but i recommend adding a `ACTION REQUIRED:` prefix to the release note.
the boostrap credential can be bound to existing infrastructures, which was probably not a great idea in the first place, but we should visibly notify about such changes.

/lgtm
/hold 
for other reviews.

**Comment by neolit123** [comment]
if we really want to be on the safe side there should be a grace period of say 1 release while `kubeadm join` supports a knob for this that defaults to `false`.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 
/hold cancel 

Given that this is a security hole I'd be ok with a backport of this. 
I don't think an action is required on the release note.

**Comment by neolit123** [comment]
> Given that this is a security hole I'd be ok with a backport of this.

i'm -1 given the workaround is very simple:
`kubeadm join .... & sudo rm /etc/kubernetes/kubelet-bootstrap.conf`

> I don't think an action is required on the release note.

i worry about somebody is copying or doing something with the bootstrap credentials.
with kubeadm being GA we should avoid such changes without notification.

**Comment by fabriziopandini** [comment]
AFAIK there is no risk in backporting this fix
- if users are using token discovery, then bootstrap-kubelet can’t be used because it expires
- if users are using file discovery, there is no way to get other credentials in until https://github.com/kubernetes/kubernetes/pull/80675 merge

But I leave the final call on this on @timothysc / @neolit123

**Review by yagonobre — COMMENTED** [review]
LGTM

**Comment by stealthybox** [comment]
the delete should be moved to defer here:
https://github.com/kubernetes/kubernetes/pull/80676/files#diff-3c566a0c1e3b213a34b39503abf4fc11R104

otherwise it's possible for the token to be left behind when the function errors

**Comment by neolit123** [comment]
yep, +1 to use defer.

**Pull request merged**
